zio-http/src/main/scala/zio/http/netty/server/NettyServerRequest.scala

```diff
@@ -0,0 +1,385 @@
+package zio.http.netty.server
+
+import io.netty.channel.ChannelHandlerContext
+import io.netty.handler.codec.http.{HttpRequest, _}
+import zio.Trace
+import zio.http._
+import zio.http.model.Method._
+import zio.http.model._
+import zio.http.netty._
+
+import java.net.{InetAddress, InetSocketAddress}
+
+private[zio] sealed abstract class NettyServerRequest(
+  val ctx: ChannelHandlerContext,
+  val nettyReq: HttpRequest,
+  currentHeaders: Option[Headers],
+  currentUrl: Option[URL],
+  currentVersion: Option[Version],
+) extends Request { self: NettyServerRequest.Update =>
+
+  override def remoteAddress: Option[InetAddress] =
+    ctx.channel().remoteAddress() match {
+      case address: InetSocketAddress => Some(address.getAddress)
+      case _                          => None
+    }
+
+  override def headers: Headers = currentHeaders.getOrElse(Headers.make(nettyReq.headers()))
+
+  override def path: Path = url.path
+
+  override def updateUrl(newUrl: URL): Request = withUpdatedUrl(newUrl)
+
+  override def updateVersion(version: Version): Request = withUpdatedVersion(version)
+
+  override def updateHeaders(update: Headers => Headers)(implicit trace: Trace): Request = withUpdatedHeaders(
+    update(headers),
+  )
+
+  override def updateMethod(newMethod: Method): Request = newMethod match {
+    case TRACE        => NettyServerRequest.trace(ctx, nettyReq, currentHeaders, currentUrl, currentVersion)
+    case POST         => NettyServerRequest.post(ctx, nettyReq, currentHeaders, currentUrl, currentVersion)
+    case HEAD         => NettyServerRequest.head(ctx, nettyReq, currentHeaders, currentUrl, currentVersion)
+    case CUSTOM(name) => NettyServerRequest.custom(ctx, name, nettyReq, currentHeaders, currentUrl, currentVersion)
+    case CONNECT      => NettyServerRequest.connect(ctx, nettyReq, currentHeaders, currentUrl, currentVersion)
+    case DELETE       => NettyServerRequest.delete(ctx, nettyReq, currentHeaders, currentUrl, currentVersion)
+    case PUT          => NettyServerRequest.put(ctx, nettyReq, currentHeaders, currentUrl, currentVersion)
+    case GET          => NettyServerRequest.get(ctx, nettyReq, currentHeaders, currentUrl, currentVersion)
+    case OPTIONS      => NettyServerRequest.options(ctx, nettyReq, currentHeaders, currentUrl, currentVersion)
+    case PATCH        => NettyServerRequest.patch(ctx, nettyReq, currentHeaders, currentUrl, currentVersion)
+  }
+
+  override def url: URL = currentUrl.getOrElse(URL.fromString(nettyReq.uri()).getOrElse(URL.empty))
+
+  override def version: Version = currentVersion.getOrElse {
```

Comment:
`getOrElse` and friends should be avoided as they take by-name parameters which allocate `Function0`.